File: prob1.java
// Time Complexity : O(2*n + max)
// Space Complexity : O(2*max)
// Did this code successfully run on Leetcode : Yes


class Solution {
    public int deleteAndEarn(int[] nums) {
        int max=0;
        for(int x:nums)
            max = Math.max(max,x);
        int[] a = new int[max+1];
        
        for(int x:nums)
            a[x]++;
        Integer[] dp = new Integer[max+1];
        return solve(a,1,dp);
    }

    private int solve(int[] a,int i,Integer[] dp){
        if(i>=a.length){
            return 0;
        }
        if(dp[i]!=null){
            return dp[i];
        }
        int pick = a[i]*i + solve(a,i+2,dp);
        int notPick = solve(a,i+1,dp);

        return dp[i] = Math.max(pick,notPick);
    }
}

File: Prob1.py
# Time - O(n)+O(max(n))
# Space - O(max(n))
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if len(nums)==0: 
            return 0        
        arr=[0 for _ in range(max(nums)+1)] #Space - O(max(n))
        for num in nums: #Time - O(n) #array of sum of duplicates. Now, this is a House of robbers problem.
            arr[num]+=num
        prev=arr[0]
        curr=max(arr[0],arr[1])
        for i in range(2,len(arr)): #Time - O(max(n))
            temp=curr
            curr=max(arr[i]+prev,curr)
            prev=temp
        return curr
        